libs/openFrameworks/video/ofAVFoundationVideoPlayer.m

```diff
@@ -29,551 +29,589 @@ @implementation ofAVFoundationVideoPlayer
+	if(self.asset == nil) {
+		NSLog(@"error loading asset: %@", [url description]);
+		return NO;
+	}
+	
+	dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+	dispatch_queue_t queue;
+	if(async){
+		queue = dispatch_get_main_queue();
+	} else {
+		queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+	}
+	
+	dispatch_async(queue, ^{
+		[self.asset loadValuesAsynchronouslyForKeys:[NSArray arrayWithObject:kTracksKey] completionHandler:^{
+			NSError * error = nil;
+			AVKeyValueStatus status = [self.asset statusOfValueForKey:kTracksKey error:&error];
+			
+			if(status != AVKeyValueStatusLoaded) {
+				NSLog(@"error loading asset tracks: %@", [error localizedDescription]);
+				if(!async){
+					dispatch_semaphore_signal(sema);
+				}
+				return;
+			}
+			
+			duration = [self.asset duration];
+			
+			if(CMTimeCompare(duration, kCMTimeZero) == 0) {
+				NSLog(@"track loaded with zero duration.");
+				if(!async){
+					dispatch_semaphore_signal(sema);
+				}
+				return;
+			}
+			
+			if(isfinite([self getDurationInSec]) == NO) {
+				NSLog(@"track loaded with infinite duration.");
+				if(!async){
+					dispatch_semaphore_signal(sema);
+				}
+				return;
+			}
+			
+			BOOL bOk = [self createAssetReaderWithTimeRange:CMTimeRangeMake(kCMTimeZero, duration)];
+			if(bOk == NO) {
+				NSLog(@"problem with creating asset reader.");
+				if(!async){
+					dispatch_semaphore_signal(sema);
+				}
+				return;
+			}
+			
+			NSArray * videoTracks = [self.asset tracksWithMediaType:AVMediaTypeVideo];
+			if([videoTracks count] == 0) {
+				NSLog(@"no video tracks found.");
+				if(!async){
+					dispatch_semaphore_signal(sema);
+				}
+				return;
+			}
+			
+			AVAssetTrack * track = [videoTracks objectAtIndex:0];
+			frameRate = track.nominalFrameRate;
+			videoWidth = [track naturalSize].width;
+			videoHeight = [track naturalSize].height;
+			
+			NSLog(@"video loaded at %i x %i", videoWidth, videoHeight);
+			
+			[self setVideoSize:CGSizeMake(videoWidth, videoHeight)];
+			
+			//------------------------------------------------------------ create player item.
+			self.playerItem = [AVPlayerItem playerItemWithAsset:self.asset];
+			
+			[self.playerItem addObserver:self
+							  forKeyPath:kStatusKey
+								 options:0
+								 context:&ItemStatusContext];
+			
+			NSNotificationCenter * notificationCenter = [NSNotificationCenter defaultCenter];
+			[notificationCenter addObserver:self
+								   selector:@selector(playerItemDidReachEnd)
+									   name:AVPlayerItemDidPlayToEndTimeNotification
+									 object:self.playerItem];
+			
+			[_player replaceCurrentItemWithPlayerItem:self.playerItem];
+			
+			[self addTimeObserverToPlayer];
+			
+			bLoaded = true;
+			
+			if(!async){
+				dispatch_semaphore_signal(sema);
+			}
+		}];
+	});
+	
+	// Wait for the dispatch semaphore signal
+	if(!async){
+		dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+		dispatch_release(sema);
```

Comment:
hey @HalfdanJ, thanks for putting this together.
my GCD knowledge is not that amazing, so just going through your code, reading up on it and working out what it all does...
have a question, since you've created a `dispatch_semaphore_t ` above, should you not release it regardless if its loading async or non-async?